buildsystem/generator/make.go

```diff
@@ -0,0 +1,300 @@
+func (m *Make) MergeVariantParameters(mvps map[string]MakeVariantParameters) *Make {
+	for name, newMVP := range mvps {
+		mv := m.Variants[name]
+		mv.MakeVariantParameters = newMVP
+		m.Variants[name] = mv
+	}
+	return m
+}
+
+// MergeEnvironments merges the given environments with the existing environment
+// variables. Duplicates are overwritten in the order in which environments are
+// passed into the function.
+func (m *Make) MergeEnvironments(envs ...map[string]string) *Make {
+	m.Environment = MergeEnvironments(append([]map[string]string{m.Environment}, envs...)...)
+	return m
+}
+
+func (m *Make) Validate() error {
+	catcher := grip.NewBasicCatcher()
+	catcher.Wrap(m.validateTasks(), "invalid task definitions")
+	catcher.Wrap(m.validateVariants(), "invalid variant definitions")
+	return catcher.Resolve()
+}
+
+func (m *Make) validateTasks() error {
+	catcher := grip.NewBasicCatcher()
+	for name, mt := range m.Tasks {
+		catcher.Wrapf(mt.Validate(), "invalid task '%s'", name)
+	}
+	return catcher.Resolve()
+}
+
+func (m *Make) validateVariants() error {
+	catcher := grip.NewBasicCatcher()
+	catcher.NewWhen(len(m.Variants) == 0, "must have at least one variant")
+	for variantName, mv := range m.Variants {
+		catcher.Wrapf(mv.Validate(), "invalid definitions for variant '%s'", variantName)
+
+		taskNames := map[string]struct{}{}
+		for _, mvt := range mv.Tasks {
+			res, err := m.getTasksAndRef(mvt)
+			if err != nil {
+				catcher.Wrapf(err, "invalid task reference in variant '%s'", variantName)
+				continue
+			}
+			for taskName := range res.tasks {
+				if _, ok := taskNames[taskName]; ok {
+					catcher.Errorf("duplicate reference to task name '%s' in variant '%s'", taskName, variantName)
+				}
+				taskNames[taskName] = struct{}{}
+			}
+		}
+	}
+	return catcher.Resolve()
+}
+
+type taskReferenceResult struct {
+	tasks map[string]MakeTask
+	ref   string
+}
+
+func (m *Make) getTasksAndRef(mvt MakeVariantTask) (*taskReferenceResult, error) {
+	if mvt.Tag != "" {
+		tasks := m.GetTasksByTag(mvt.Tag)
+		if len(tasks) == 0 {
+			return nil, errors.Errorf("no tasks matched tag '%s'", mvt.Tag)
+		}
+		return &taskReferenceResult{tasks: tasks, ref: mvt.Tag}, nil
+	}
+
+	if mvt.Name != "" {
+		task, err := m.GetTaskByName(mvt.Name)
+		if err != nil {
+			return nil, errors.Wrapf(err, "finding definition for package named '%s'", mvt.Name)
+		}
+		tasks := map[string]MakeTask{mvt.Name: *task}
+		return &taskReferenceResult{tasks: tasks, ref: mvt.Name}, nil
+	}
+
+	return nil, errors.New("empty package reference")
+}
+
+func (m *Make) Generate() (*shrub.Configuration, error) {
+	conf, err := shrub.BuildConfiguration(func(c *shrub.Configuration) {
+		for vName, mv := range m.Variants {
+			variant := c.Variant(vName)
+			variant.DistroRunOn = mv.Distros
+
+			var tasksForVariant []*shrub.Task
+			for _, mvt := range mv.Tasks {
+				res, err := m.getTasksAndRef(mvt)
+				if err != nil {
+					panic(err)
```

Comment:
just to check, but the BuildConfiguration handles panics?

Reply:
Yeah, it's the only way it can produce errors.